File: CS/MasterDetailSelectionHelper.cs
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace TestMasterDetailSelection
{
    public class MasterDetailSelectionHelper
    {
        public MasterDetailSelectionHelper(GridView gridView)
        {
            _GridView = gridView;
            _GridView.MasterRowExpanded += _GridView_MasterRowExpanded;
            _GridView.SelectionChanged += _GridView_SelectionChanged;
        }

        void _GridView_SelectionChanged(object sender, DevExpress.Data.SelectionChangedEventArgs e)
        {
            if (_GridView.IsDetailView)
            {
                int masterRowHandle = _GridView.SourceRowHandle;
                if (_GridView.GetSelectedRows().Length == _GridView.RowCount)
                {
                    (_GridView.ParentView as GridView).SelectRow(masterRowHandle);
                }
                else if (_GridView.GetSelectedRows().Length == 0)
                    (_GridView.ParentView as GridView).UnselectRow(masterRowHandle);
            }
            if (_GridView.IsMasterRow(e.ControllerRow) || e.ControllerRow == GridControl.InvalidRowHandle)
            {
                UpdateSelection(e.ControllerRow);
            }
            if (e.Action == CollectionChangeAction.Refresh)
            {
                foreach (BaseView view in _GridView.GridControl.Views)
                {
                    if (view.IsDetailView)
                    {
                        int masterRowHandle = view.SourceRowHandle;
                        UpdateSelection(masterRowHandle);
                    }
                }
            }
        }

        void _GridView_MasterRowExpanded(object sender, CustomMasterRowEventArgs e)
        {
            UpdateSelection(e.RowHandle);
        }

        void UpdateSelection(int masterRowHandle)
        {
            GridView detailView = _GridView.GetDetailView(masterRowHandle, 0) as GridView;
            if (detailView != null)
            {
                if (_GridView.IsRowSelected(masterRowHandle))
                {
                    detailView.SelectAll();
                }
                else
                {
                    detailView.ClearSelection();
                }
            }

        }

        GridView _GridView;
    }
}

File: VB/MasterDetailSelectionHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace TestMasterDetailSelection
	Public Class MasterDetailSelectionHelper
		Public Sub New(ByVal gridView As GridView)
			_GridView = gridView
			AddHandler _GridView.MasterRowExpanded, AddressOf _GridView_MasterRowExpanded
			AddHandler _GridView.SelectionChanged, AddressOf _GridView_SelectionChanged
		End Sub

		Private Sub _GridView_SelectionChanged(ByVal sender As Object, ByVal e As DevExpress.Data.SelectionChangedEventArgs)
			If _GridView.IsDetailView Then
				Dim masterRowHandle As Integer = _GridView.SourceRowHandle
				If _GridView.GetSelectedRows().Length = _GridView.RowCount Then
					TryCast(_GridView.ParentView, GridView).SelectRow(masterRowHandle)
				ElseIf _GridView.GetSelectedRows().Length = 0 Then
					TryCast(_GridView.ParentView, GridView).UnselectRow(masterRowHandle)
				End If
			End If
			If _GridView.IsMasterRow(e.ControllerRow) OrElse e.ControllerRow = GridControl.InvalidRowHandle Then
				UpdateSelection(e.ControllerRow)
			End If
			If e.Action = CollectionChangeAction.Refresh Then
				For Each view As BaseView In _GridView.GridControl.Views
					If view.IsDetailView Then
						Dim masterRowHandle As Integer = view.SourceRowHandle
						UpdateSelection(masterRowHandle)
					End If
				Next view
			End If
		End Sub

		Private Sub _GridView_MasterRowExpanded(ByVal sender As Object, ByVal e As CustomMasterRowEventArgs)
			UpdateSelection(e.RowHandle)
		End Sub

		Private Sub UpdateSelection(ByVal masterRowHandle As Integer)
			Dim detailView As GridView = TryCast(_GridView.GetDetailView(masterRowHandle, 0), GridView)
			If detailView IsNot Nothing Then
				If _GridView.IsRowSelected(masterRowHandle) Then
					detailView.SelectAll()
				Else
					detailView.ClearSelection()
				End If
			End If

		End Sub

		Private _GridView As GridView
	End Class
End Namespace
